### PR TITLE
[WIP][SPARK-42578][CONNECT] Add JDBC to DataFrameWriter

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
Currently, the connect project have the new `DataFrameWriter` API which is corresponding to Spark `DataFrameWriter` API. But the connect's `DataFrameWriter` missing the jdbc API.

### Why are the changes needed?
This PR try to add JDBC to `DataFrameWriter`.

### Does this PR introduce _any_ user-facing change?
'No'.
New feature.

### How was this patch tested?
@hvanhovell It seems that add test cases no way.

### CODE REVIEW

**Comment by beliefer** [comment]
@hvanhovell It seems that add test cases no way.

**Comment by hvanhovell** [comment]
hmmm - let me think about it.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 1]
```diff
@@ -345,6 +345,37 @@ final class DataFrameWriter[T] private[sql] (ds: Dataset[T]) {
...
+   *   JDBC database connection arguments, a list of arbitrary string tag/value. Normally at least
+   *   a "user" and "password" property should be included. "batchsize" can be used to control the
+   *   number of rows per insert. "isolationLevel" can be one of "NONE", "READ_COMMITTED",
+   *   "READ_UNCOMMITTED", "REPEATABLE_READ", or "SERIALIZABLE", corresponding to standard
+   *   transaction isolation levels defined by JDBC's Connection object, with default of
+   *   "READ_UNCOMMITTED".
+   * @since 3.4.0
+   */
+  def jdbc(url: String, table: String, connectionProperties: Properties): Unit = {
+    // connectionProperties should override settings in extraOptions.
```
Comment (LuciferYang) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala:372]: I have a question @hvanhovell @beliefer . For the connect-client api, should we verify the parameters on the client side or on the server side?

Comment (beliefer) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala:372]: I think verify parameters on the server side is a robust way. Certainly, the work on client side will reduce the pressure on the server side.

**Comment by hvanhovell** [comment]
@beliefer we should be able to create an in-memory table and append a couple of rows to that right?

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 2]
```diff
@@ -345,6 +345,48 @@ final class DataFrameWriter[T] private[sql] (ds: Dataset[T]) {
...
+    this.extraOptions ++= connectionProperties.asScala
+    // explicit url and dbtable should override all
+    this.extraOptions ++= Seq("url" -> url, "dbtable" -> table)
+    format("jdbc").saveAsJdbcTable(table)
+  }
+
+  private def saveAsJdbcTable(tableName: String): Unit = {
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala:379]: Why do we need a separate method? There is only one method using it.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 3]
```diff
@@ -345,6 +345,44 @@ final class DataFrameWriter[T] private[sql] (ds: Dataset[T]) {
...
+   * @param connectionProperties
+   *   JDBC database connection arguments, a list of arbitrary string tag/value. Normally at least
+   *   a "user" and "password" property should be included. "batchsize" can be used to control the
+   *   number of rows per insert. "isolationLevel" can be one of "NONE", "READ_COMMITTED",
+   *   "READ_UNCOMMITTED", "REPEATABLE_READ", or "SERIALIZABLE", corresponding to standard
+   *   transaction isolation levels defined by JDBC's Connection object, with default of
+   *   "READ_UNCOMMITTED".
+   * @since 3.4.0
+   */
+  def jdbc(url: String, table: String, connectionProperties: Properties): Unit = {
```
Comment (LuciferYang) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala:371]: I think we can remove `ProblemFilters.exclude[Problem]("org.apache.spark.sql.DataFrameWriter.jdbc")` from `CheckConnectJvmClientCompatibility` in this pr
`

Comment (beliefer) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala:371]: Thank you for the reminder.

File: connector/connect/server/pom.xml [hunk 4]
```diff
@@ -199,6 +199,12 @@
       <version>${tomcat.annotations.api.version}</version>
       <scope>provided</scope>
     </dependency>
+    <dependency>
```
Comment (hvanhovell) [inline connector/connect/server/pom.xml:202]: Why the move?

**Review by hvanhovell — COMMENTED** [review]
Looks good! Can you look into the test failure?

File: connector/connect/server/pom.xml (hunk 4 above)
Comment (beliefer) [reply connector/connect/server/pom.xml:207]: It seems the server can't find the jar from class path.
image

**Comment by hvanhovell** [comment]
@beliefer are you abandoning this one?

**Comment by beliefer** [comment]
> @beliefer are you abandoning this one?

Because other PR implement this function.

**Comment by hvanhovell** [comment]
Is that https://github.com/apache/spark/pull/40415?

**Comment by beliefer** [comment]
> Is that #40415?

It is https://github.com/apache/spark/pull/40358

**Pull request closed without merging**